GitVersionCore/GitVersionContext.cs

```diff
@@ -81,30 +81,83 @@ IEnumerable<Branch> GetBranchesContainingCommit(string commitSha)
 
         void CalculateEffectiveConfiguration()
         {
-            var matchingBranches = configuration.Branches.Where(b => Regex.IsMatch("^" + CurrentBranch.Name, b.Key)).ToArray();
+            var currentBranchConfig = GetBranchConfiguration(CurrentBranch);
 
-            var currentBranchConfig = GetBranchConfiguration(matchingBranches);
-
-            var versioningMode = currentBranchConfig.VersioningMode ?? configuration.VersioningMode ?? VersioningMode.ContinuousDelivery;
-            var tag = currentBranchConfig.Tag;
+            var versioningMode = currentBranchConfig.Value.VersioningMode ?? configuration.VersioningMode ?? VersioningMode.ContinuousDelivery;
+            var tag = currentBranchConfig.Value.Tag;
             var nextVersion = configuration.NextVersion;
-
-            Configuration = new EffectiveConfiguration(configuration.AssemblyVersioningScheme, versioningMode, configuration.TagPrefix, tag, nextVersion);
+            var incrementStrategy = currentBranchConfig.Value.Increment ?? IncrementStrategy.Patch;
+            var assemblyVersioningScheme = configuration.AssemblyVersioningScheme;
+            var gitTagPrefix = configuration.TagPrefix;
+            Configuration = new EffectiveConfiguration(assemblyVersioningScheme, versioningMode, gitTagPrefix, tag, nextVersion, incrementStrategy, currentBranchConfig.Key);
         }
 
-        BranchConfig GetBranchConfiguration(KeyValuePair<string, BranchConfig>[] matchingBranches)
+        KeyValuePair<string, BranchConfig> GetBranchConfiguration(Branch currentBranch)
```

Comment:
Possible, but we already have a Dictionary so didn't bother converting between. Also was internal to the class so was not too concerned